terraform/aws/irsa.tf

```diff
@@ -1,20 +1,38 @@
-data "aws_caller_identity" "current" {}
+/*
+  This file provides resources _per hub and role_. Each role is tied to a
+  specific k8s ServiceAccount allowed to assume the role.
+
+  - Role                 - for use by k8s ServiceAccount (user-sa, admin-sa)
+  - Policy               - if extra_iam_policy is declared
+  - RolePolicyAttachment - if extra_iam_policy is declared
+*/
 
+data "aws_caller_identity" "current" {}
 data "aws_partition" "current" {}
 
-resource "aws_iam_role" "irsa_role" {
-  for_each = var.hub_cloud_permissions
-  name     = "${var.cluster_name}-${each.key}"
 
-  assume_role_policy = data.aws_iam_policy_document.irsa_role_assume[each.key].json
+
+locals {
+  hub_role = flatten([
+    for hub, hub_value in var.hub_cloud_permissions : [
+      for role, role_value in hub_value : {
```

Comment:
My understanding is that the *key* here refers to the name of the Kubernetes Service Account (because that is what is specified in the variable). The name of the kubernetes service account is the same as the name of the IAM role (with a prefix) in some cases (`user-sa`) but not others. And `id` is here our own creation, and not the same as the `id` of the `iam_role` resource (https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role#id).

I would suggest the following fields:

1. `iam_role_name` (currently called `id`)
2. `ksa_name` (currently called `role`)

I believe this would make this intermediate data structure easier to understand when being called.

Reply:
This would make things like https://github.com/2i2c-org/infrastructure/pull/3932/files#diff-9f223d043fe1fe9d6f6e3f3a909a3e8d633b2fbbc20f388af72c20069d53cc5aR49 much more obvious for example - we're using that in the context of the kubernetes service account name inside a namespace, not the name of the role.

Reply:
Thanks for thinking about naming, I agree fully that these are improvements and I think in this case - quite important improvements for anyone trying to understand this code.